File: solutions/JavaScript/design_search_words_recursive.js
class TrieNode {
    constructor() {
        this.children = {};
        this.isEndOfWord = false;
    }
}

class WordDictionary {
    constructor() {
        this.root = new TrieNode();
    }

    addWord(word) {
        let node = this.root;
        for (let char of word) {
            if (!(char in node.children)) {
                node.children[char] = new TrieNode();
            }
            node = node.children[char];
        }
        node.isEndOfWord = true;
    }

    search(word) {
        return this.searchDFS(word, 0, this.root);
    }

    searchDFS(word, index, node) {
        if (index === word.length) {
            return node.isEndOfWord;
        }

        let char = word[index];
        if (char === '.') {
            for (let child of Object.values(node.children)) {
                if (this.searchDFS(word, index + 1, child)) {
                    return true;
                }
            }
            return false;
        } else {
            if (!(char in node.children)) {
                return false;
            }
            return this.searchDFS(word, index + 1, node.children[char]);
        }
    }
}

// Example usage
const wd = new WordDictionary();
wd.addWord("bad");
wd.addWord("dad");
wd.addWord("mad");
console.log(wd.search("pad"));  // false
console.log(wd.search("bad"));  // true
console.log(wd.search(".ad"));  // true
console.log(wd.search("b.."));  // true

File: solutions/Python/design_search_words_recursive.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_end_of_word = False


class WordDictionary:
    def __init__(self):
        self.root = TrieNode()

    def addWord(self, word: str) -> None:
        node = self.root
        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]
        node.is_end_of_word = True

    def search(self, word: str) -> bool:
        return self._search_dfs(word, 0, self.root)

    def _search_dfs(self, word: str, index: int, node: TrieNode) -> bool:
        if index == len(word):
            return node.is_end_of_word

        char = word[index]
        if char == '.':
            for child in node.children.values():
                if self._search_dfs(word, index + 1, child):
                    return True
            return False
        else:
            if char not in node.children:
                return False
            return self._search_dfs(word, index + 1, node.children[char])


# Example usage
wd = WordDictionary()
wd.addWord("bad")
wd.addWord("dad")
wd.addWord("mad")
print(wd.search("pad"))  # False
print(wd.search("bad"))  # True
print(wd.search(".ad"))  # True
print(wd.search("b.."))  # True
